engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/TypePropagation.java

```diff
@@ -164,6 +175,45 @@ TypeRepresentation tryInferringType(
     return ascribedType != null ? ascribedType : inferredType;
   }
 
+  private boolean canBeDiscarded(TypeRepresentation type) {
+    if (type == null || type instanceof TypeRepresentation.TopType) {
+      // If the type is unknown then we allow it to be discarded as we don't know enough yet
+      return true;
+    }
+
+    if (type.equals(BuiltinTypes.NOTHING)) {
+      // Nothing is the type that side-effectful functions should return - it is most often meant to
+      // be discarded.
+      return true;
+    }
+
+    // Sometimes Nothing is inferred as `type Nothing`. Maybe we should fix it, but for now:
+    if (type instanceof TypeRepresentation.TypeObject typeObject) {
+      if (typeObject.instanceType().equals(BuiltinTypes.NOTHING)) {
+        return true;
+      }
+    }
+
+    if (type instanceof TypeRepresentation.ArrowType) {
+      // Under all scenarios, discarding a not-fully applied function is an error - such an
+      // operation is essentially a no-op.
+      // (in fact there are exceptions due to currying, but in 99% cases this is an error)
+      return false;
+    }
+
+    return canNonFunctionsBeDiscarded;
+  }
+
+  /**
+   * Drives the `canBeDiscarded` method.
+   *
+   * <ol>
+   *   <li>If true, then only a not-fully-applied function will raise discarded value warnings.
+   *   <li>If false, any non-Nothing value will raise a warning.
+   * </ol>
+   */
+  protected boolean canNonFunctionsBeDiscarded = true;
```

Comment:
Once the check matures we can remove this property and simplify the `canBeDiscarded` method.

For now I wanted to keep it to relatively easily allow switching between 'warn on Functions only' and 'warn on any non-Nothing value', to see which setting makes more sense. It is so internal it didn't feel that it was worth to expose as a CLI flag though.

I will use this to do some testing once the CI Lint pass is implemented and then will remove (settling on one of the two modes).